Develop a ComponentDefinition for a cloud resource (#261)
oam-dev/kubevela.io: docs/platform-engineers/components/component-terraform.md

$ kubectl apply -f https://raw.githubusercontent.com/oam-dev/terraform-controller/master/examples/alibaba/provider.yaml
provider.terraform.core.oam.dev/default created
```

## Develop a ComponentDefinition for a cloud resource

### Alibaba Cloud

> Take [Elastic IP](https://www.alibabacloud.com/help/doc-detail/36016.htm) as an example.

Reviewer: ```suggestion
In the following we will develop a Terraform resource for [Elastic IP](https://www.alibabacloud.com/help/doc-detail/36016.htm).
```

> - Develop a Terraform resource or module for the cloud resource

Reviewer: ```suggestion
```

> #### Develop a Terraform resource or module for the cloud resource

Reviewer 2: 洪超是对的，这个小节可以删掉，你觉得有用，就只保留一句话:

> You can search the required cloud resource module from [Terraform official module registry](https://registry.terraform.io/browse/modules).

Author: @wonderflow @hongchaodeng Delete this section. Thanks for the suggestion.

> We recommend you to search the required cloud resource module in [Terraform official module registry](https://registry.terraform.io/browse/modules).

Reviewer: 这两段都删掉。。作为最后的 Appendix 即可

> If no luck, you can create a Terraform resource or module yourself per

Reviewer: 删掉

Author: 这个其实很重要，是不是觉得太琐碎了？

Reviewer: 放在这里干啥？不要放跟上下文没有直接关系的。
认为重要的放 Appendix

Author: Deleted. Thanks.

[Terraform Alibaba Cloud Provider specifications](https://registry.terraform.io/providers/aliyun/alicloud/latest/docs).

```terraform
module "eip" {
  source = "github.com/zzxwill/terraform-alicloud-eip"
  name = var.name
  bandwidth = var.bandwidth
}

variable "name" {
  description = "Name to be used on all resources as prefix. Default to 'TF-Module-EIP'."
  default = "TF-Module-EIP"
  type = string
}

variable "bandwidth" {
  description = "Maximum bandwidth to the elastic public network, measured in Mbps (Mega bit per second)."
  type = number
  default = 5
}

output "EIP_ADDRESS" {
  description = "The elastic ip address."
  value       = module.eip.this_eip_address.0
}
```

> - Develop a ComponentDefinition for the cloud resource

Reviewer: ```suggestion
Then develop a ComponentDefinition for the EIP resource:
```

Author: It's a title, I don't want it to be `EIP` specified, just taking EIP as an example.

Reviewer: 那就把格式改一改吧

Author: Done.

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: ComponentDefinition
metadata:
  name: alibaba-eip
  namespace: {{.Values.systemDefinitionNamespace}}
  annotations:
    definition.oam.dev/description: Terraform configuration for Alibaba Cloud Elastic IP
  labels:
    type: terraform
spec:
  workload:
    definition:
      apiVersion: terraform.core.oam.dev/v1beta1
      kind: Configuration
  schematic:
    terraform:
      configuration: |
        module "eip" {
          source = "github.com/zzxwill/terraform-alicloud-eip"
          name = var.name
          bandwidth = var.bandwidth
        }

        variable "name" {
          description = "Name to be used on all resources as prefix. Default to 'TF-Module-EIP'."
          default = "TF-Module-EIP"
          type = string
        }

        variable "bandwidth" {
          description = "Maximum bandwidth to the elastic public network, measured in Mbps (Mega bit per second)."
          type = number
          default = 5
        }

        output "EIP_ADDRESS" {
          description = "The elastic ip address."
          value       = module.eip.this_eip_address.0
        }

```

> - Verify

Reviewer: 这种格式还是有很大的问题。尽量不要用这种列表，因为每个点都太长了。如果要强调就改成大标题。

Author: 好。
